Group/ungroup

petehunt opened the pull request:
Adds keynote-style group/ungroup to excalidraw. Putting it up so people can vet the approach and start to play with it. Is not thoroughly tested yet.

Play with it by selecting a group of elements and pressing cmd-option-g to group, and shift-cmd-option-g to ungroup. You should be able to have nested groups without much trouble as well.

Code definitely needs to be cleaned up too, and tests need to be written.

I spoke w/ @vjeux and decided to make groups a new concept rather than create a new type of ExcalidrawElement to represent them. This made the implementation a lot easier, but it also means that code dealing with selection needs to think about group and non-group cases. I think this can be solved with some better abstraction around the selection state.

dai-shi commented:
Nice! 🎉 
(Here's my secret: I was hoping to use group information for animation.)

dwelle commented:
Super cool. Will check code later.

> Play with it by selecting a group of elements and pressing cmd-option-g to group, and shift-cmd-option-g to ungroup.

Why not go with standard `Cmd+G` & `Cmd+Shift+G`?

dai-shi commented:
Quickly played with it and found this.
(Double clicking to select a group element was so intuitive!)

![exdraw45](https://user-images.githubusercontent.com/490574/82810212-dc9ac600-9ec8-11ea-920b-d05161a72854.gif)

petehunt commented:
For this feature I basically tried to emulate keynote which uses cmd option g. Definitely down to change it though. 

Yep, didn’t test rotation, will fix :)

dwelle commented:
I think the implementation is the right way to go.

Here's one bug I found:

1. create 2 elements and group them
2. duplicate the group
3. select both groups, and duplicate it again
4. group 2 & 3 are grouped together (well technically, 2 & 3 are not groups, but a group of 4 elements now)

![excal_group_bug](https://user-images.githubusercontent.com/5153846/82820485-8dda3580-9ea2-11ea-99db-4b25cf8465d5.gif)

j-f1 commented:
Can you add group and ungroup buttons to the “Actions” section of the shape actions menu?

lipis commented:
Also shortcuts.. I guess `Cmd+G`, `Cmd+U`

j-f1 commented:
bug:
![QuickTime movie](https://user-images.githubusercontent.com/25517624/82834776-1b605a00-9e90-11ea-97ab-df514bc45170.gif)

Looks great otherwise!

j-f1 commented:
Another related issue: you can select an object in a group and group it with an object outside of the group. It seems like Keynote handles this by deselecting everything when you click something that’s not a group object if you’re selecting individual objects in a group.

j-f1 commented:
Interesting — Keynote seems to ignore everything that’s not in the current group for selection, so clicking an object that’s not in the group is equivalent to clicking on the scene background.

petehunt commented:
ok, this is ready for code review. i think i fixed all the bugs called out in the thread (plus some more that i found as i was testing), but some may have snuck in in the last commit where i refactored a bit. will continue to test

dai-shi commented:
Is it expected that grouping doesn't take into account layers?

![exdraw48](https://user-images.githubusercontent.com/490574/82852062-ba429000-9f3c-11ea-8fa2-2d953c7db2b9.gif)

petehunt commented:
yes, that's expected. i'd suggest we do it it in a follow up...

lipis commented:
> Is it expected that grouping doesn't take into account layers?

Let's file an issue for it so we won't forget.

lipis commented:
Also let's add the shortcuts to the shortcuts dialog in this PR.

dwelle commented on src/actions/actionGroup.ts:
```diff
@@ -0,0 +1,119 @@
+    return (
+      !event.shiftKey &&
+      event[KEYS.CTRL_OR_CMD] &&
+      event.keyCode === KEYS.G_KEY_CODE
```
I believe `keyCode` was used because previously `event.altKey` was messing up the combination on Mac. This Since we're no longer using `altKey`, we should use `event.key` to be layout-agnostic.

Can someone test this on on Mac?

dwelle replied:
I've pushed the fix, so check preview. Will revert if it doesn't work on Mac.

petehunt replied:
this needs to be revered. shift-cmd-g does not work

dwelle replied:
Ok. Hopefully most layouts have `G` in the right place.

petehunt commented:
oh shit, good call on the migration. that would have been a doozy. thanks!

dwelle commented:
Also, feel free to revert the label change. I copied what Figma uses because I felt the shortcut dialog may use a longer label, but maybe it's better to go with minimal text.

petehunt commented:
retested with the revert, it worked
